src/Conversion/ONNXToKrnl/ControlFlow/Loop.cpp

```diff
@@ -189,7 +204,56 @@ struct ONNXLoopOpLowering : public ConversionPattern {
           thenBlock.end(), loopBodyBlock.getOperations());
       rewriter.eraseBlock(&loopBodyBlock);
     }
-    rewriter.replaceOp(op, outputs);
+
+    rewriter.restoreInsertionPoint(afterLoop);
+    // accumulate dynamic tensor
+    // Convert the memref<?xmemref<>> to a memref
+    SmallVector<Value, 4> newOutputs;
+    for (auto output : outputs) {
+      auto seqElementType =
+          output.getType().cast<MemRefType>().getElementType();
+      if (seqElementType.isa<MemRefType>()) {
+        // need convertion
+        // ToDo: need a IF statement to handle output is empty
+        // we can safely give 0 to the dynamic dim for alloc
+        // Here loop is assumed to be executed at least once.
+        auto firstElement =
+            create.krnl.load(output, create.math.constantIndex(0));
+        SmallVector<mlir::Value, 4> allocParams;
+        SmallVector<int64_t, 4> dims;
+        dims.emplace_back(output.getType().cast<MemRefType>().getShape()[0]);
+        if (output.getType().cast<MemRefType>().getShape()[0] == -1)
+          allocParams.emplace_back(create.mem.dim(output, 0));
+        for (auto i = 0;
+             i < firstElement.getType().cast<MemRefType>().getRank(); i++) {
+          dims.emplace_back(
+              firstElement.getType().cast<MemRefType>().getShape()[i]);
+          if (firstElement.getType().cast<MemRefType>().getShape()[i] == -1)
+            allocParams.emplace_back(create.mem.dim(firstElement, i));
+        }
+        ArrayRef<int64_t> shape(dims.data(), dims.size());
+        auto flatType = MemRefType::get(
+            shape, firstElement.getType().cast<MemRefType>().getElementType());
+        auto alloc = create.mem.alignedAlloc(flatType, allocParams);
+        // copy the value
+        BuildKrnlLoop loop(rewriter, loc, 1);
```

Comment:
I would suggest to use the new krnl loop creations createKrnl.iterate (here and above), if not too disruptive.

Reply:
This is a copy of old code, which uses an utility function, emitCopy, to generate code.  Update this code when emitCopy is updated to use builder instead of rewriter.